### PR TITLE
new input data structure closing #674

### PR DESCRIPTION
**Pull request by pmrv**
there's a few open issues, that I'd like to discuss first
1. storing the list explicit vs. pickling; @jan-janssen gave good reasons today in the meeting not to pickle, my remaining question would then only be if we want to store version+type information in the HDF5 or just  the hierarchical data
2. I'm unsure about what happens (or should happen) on loading existing jobs; do we want to load old jobs with GenericParamaters, should it be automatically converted, etc.
3. I haven't added unit tests yet, only the original doctests, but I'll convert those soon-ish

I added a change keep all sphinx input groups in a single `Group()` as we discussed in the original issue, but that breaks a bunch of tests I haven't fixed yet.

### CODE REVIEW

**Comment by jan-janssen** [comment]
The HDF5 format supports storing JSON data, this still requires loading the whole JSON struggle when we want to access one entry, but having many small nodes in the HDF5 file is also bad for compression - therefore my suggestion storing the whole object as JSON https://github.com/pyiron/pyiron/blob/master/pyiron/base/generic/hdfio.py#L700

**Comment by pmrv** [comment]
I think json is a good idea, I timed deserializing a small piece of json (the k-point path of a sphinx input) and it takes about ~10µs.  So you'd have to go over 10-100 thousand jobs before this because the noticeable instead of the hdf5 access.

**Comment by pmrv** [comment]
I think this should be mostly finished, only some small things to discuss:
1. GenericParameters derives from PyironObject, but never overloads the required methods.  I could also derive InputList from there, but I'm not sure what the benefit is.
2. Codacy complains about me overriding get/update, but the new form should be consistent and I thought it's more convenient, but I could also make separate methods.
3. I'm a bit unsure about `_type_to_hdf()`. I more or less adapted it from GenericParameters, but I don't understand really what the different fields are for.  Can we document this somewhere?

All in all I touched a little bit more stuff than I wanted to, because all the initialization also changed.  I ran the sphinx handbook without errors, but if you have some of your own notebooks I would appreciate if you try this branch with them (or point me to them).

**Comment by jan-janssen** [comment]
There still seems to be an issue when executing Sphinx calculation: 
```
---------------------------------------------------------------------------
2691Exception encountered at "In [6]":
2692---------------------------------------------------------------------------
2693ValueError                                Traceback (most recent call last)
2694<ipython-input-6-effcd8a4eb8b> in <module>
2695      4 job.structure += pr.create_atoms(elements=['C'], positions=[[0, 0, 0.5*a_Fe]], magmoms=[0])
2696      5 job.calc_static()
2591----> 6 job.run()
2592
2593~/build/pyiron/pyiron/pyiron/base/job/generic.py in run(self, run_again, repair, debug, run_mode)
2594    673                 status = "created"
2595    674             if status == "initialized":
2596--> 675                 self._run_if_new(debug=debug)
2597    676             elif status == "created":
2598    677                 que_id = self._run_if_created()
2599
2600~/build/pyiron/pyiron/pyiron/base/job/generic.py in _run_if_new(self, debug)
2601   1336         else:
2602   1337             self._create_job_structure(debug=debug)
2603-> 1338             self.run()
2604   1339 
2605   1340     def _run_if_created(self):
2606
2607~/build/pyiron/pyiron/pyiron/base/job/generic.py in run(self, run_again, repair, debug, run_mode)
2608    675                 self._run_if_new(debug=debug)
2609    676             elif status == "created":
2610--> 677                 que_id = self._run_if_created()
2611    678                 if que_id:
2612    679                     self._logger.info(
2613
2614~/build/pyiron/pyiron/pyiron/base/job/generic.py in _run_if_created(self)
2615   1353             self.run_if_manually()
2616   1354         elif self.server.run_mode.modal:
2617-> 1355             self.run_static()
2618   1356         elif self.server.run_mode.non_modal or self.server.run_mode.thread:
2619   1357             self.run_if_non_modal()
2620
2621~/build/pyiron/pyiron/pyiron/base/job/generic.py in run_static(self)
2622    775             "{}, status: {}, output: {}".format(self.job_info_str, self.status, out)
2623    776         )
2624--> 777         self.run()
2625    778         if job_crashed:
2626    779             self.status.aborted = True
2627
2628~/build/pyiron/pyiron/pyiron/base/job/generic.py in run(self, run_again, repair, debug, run_mode)
2629    688                 self._run_if_running()
2630    689             elif status == "collect":
2631--> 690                 self._run_if_collect()
2632    691             elif status == "suspend":
2633    692                 self._run_if_suspended()
2634
2635~/build/pyiron/pyiron/pyiron/base/job/generic.py in _run_if_collect(self)
2636   1424         status is set to 'finished'
2637   1425         """
2638-> 1426         self.collect_output()
2639   1427         self.collect_logfiles()
2640   1428         if self.job_id is not None:
2641
2642~/build/pyiron/pyiron/pyiron/sphinx/interactive.py in collect_output(self, force_update)
2643    223 
2644    224     def collect_output(self, force_update=False):
2645--> 225         super(SphinxInteractive, self).collect_output(force_update=force_update)
2646    226         self._output_interactive_to_generic()
2647    227 
2648
2649~/build/pyiron/pyiron/pyiron/sphinx/base.py in collect_output(self, force_update)
2650   1283         Collects the outputs and stores them to the hdf file
2651   1284         """
2652-> 1285         self._output_parser.collect(directory=self.working_directory)
2653   1286         self._output_parser.to_hdf(self._hdf5, force_update=force_update)
2654   1287 
2655
2656~/build/pyiron/pyiron/pyiron/sphinx/base.py in collect(self, directory)
2657   2214             directory (str): the directory to collect the output from.
2658   2215         """
2659-> 2216         self.collect_sphinx_log(file_name="sphinx.log", cwd=directory)
2660   2217         self.collect_energy_dat(file_name="energy.dat", cwd=directory)
2661   2218         self.collect_residue_dat(file_name="residue.dat", cwd=directory)
2662
2663~/build/pyiron/pyiron/pyiron/sphinx/base.py in collect_sphinx_log(self, file_name, cwd, check_consistency)
2664   2101             if len(forces) != 0:
2665   2102                 forces = np.array(forces).reshape(
2666-> 2103                     -1, len(self._job.structure), 3)
2667   2104                 for ii, ff in enumerate(forces):
2668   2105                     forces[ii] = ff[self._job.id_spx_to_pyi]
2669
2670ValueError: cannot reshape array of size 6 into shape (3,3)
```

It should be possible to reproduce this issue on mybinder:
[Binder](https://mybinder.org/v2/gh/pyiron/pyiron/inputlist)

**Comment by jan-janssen** [comment]
The critical part is: 
```
job = pr.create_job(pr.job_type.Sphinx, 'spx_Fe_ferro_C')
job.structure = pr.create_structure('Fe', 'bcc', a_Fe)
job.structure.set_initial_magnetic_moments([2, 2])
job.structure += pr.create_atoms(elements=['C'], positions=[[0, 0, 0.5*a_Fe]], magmoms=[0])
job.calc_static()
job.run()
```

**Comment by pmrv** [comment]
The problem is that the sphinx.input.structure is not kept synchronized after the carbon is added.  Will have a detailed look.

File: pyiron/base/generic/inputlist.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
+"""
```
Comment (jan-janssen) [inline pyiron/base/generic/inputlist.py:1]: Please add the copyright and author statement

**Comment by jan-janssen** [comment]
From my side there is only one small change, but maybe @samwaseda should also take a look at these changes.

**Pull request merged**